README.md
# Puppeteer wrapper

![MIT license](https://img.shields.io/badge/license-MIT-blue.svg)
![npm version](https://img.shields.io/npm/v/@touch4it/puppeteer)
![node version](https://img.shields.io/node/v/@touch4it/puppeteer)
![vulnerabilities](https://img.shields.io/snyk/vulnerabilities/npm/@touch4it/puppeteer)

Wrapper around [Puppeteer](https://pptr.dev/) library to make it easier to use.

Puppeteer is a Node library which provides a high-level API to control Chrome or Chromium over the DevTools Protocol. Puppeteer runs headless by default, but can be configured to run full (non-headless) Chrome or Chromium.

## Installing

Install puppeteer wrapper from npm

```bash
npm i puppeteer @touch4it/puppeteer
```

or

```bash
npm i puppeteer-core @touch4it/puppeteer
```

In order to use it you will need to require it first

```javascript
const puppeteerT4IT = require('@touch4it/puppeteer');
```

## Documentation

### Get Puppeteer instance

Returns [Puppeteer](https://www.npmjs.com/package/puppeteer) library.

```javascript
const puppeteer = puppeteerT4IT.getPuppeteer();
```

### Change default TIMEOUT and SCREENSHOT_DIRECTORY vars

If you want to use different timeout and screenshot directory, you can use init function to change these variables. You can pass following arguments into this function:

* `timeout` Default request timeout which will be use in the module (defaults to `30000` ms)
* `screenshotDirectory` Name of the directory where screenshots will be stored (defaults to `screenshots/`)

```javascript
puppeteerT4IT.init(5000, 'some_other_dir/');
```

### Launch new browser instance

You can pass `option` object as argument, which can contain all available option for [launch()](https://pptr.dev/#?product=Puppeteer&version=v1.11.0&show=api-puppeteerlaunchoptions) function

Returns new [Browser](https://pptr.dev/#?product=Puppeteer&version=v1.11.0&show=api-class-browser) instance.

```javascript
let browser = await puppeteerT4IT.launch({ headless: false });
```

### Get new page instance

You can pass following arguments into this function:

* `browser` A [Browser](#Launch-new-browser-instance) instance that you have created with [puppeteer.launch()](#Launch-new-browser-instance)
* `authenticationData` Object containing "username" and "password" [authentication](https://pptr.dev/#?product=Puppeteer&version=v1.11.0&show=api-pageauthenticatecredentials) keys for basic auth when the page is launched

Returns new [Page](https://pptr.dev/#?product=Puppeteer&version=v1.11.0&show=api-class-page) instance.

```javascript
let page = await puppeteerT4IT.getPage(browser, { username: 'User', password: '123abc' });
```

### Navigate to web page and test status code

You can pass following arguments into this function:

* `page` Puppeteer [page](#Get-new-page-instance) object
* `url` URL to be redirected to
* `statusCode` awaited HTTP response status code (defaults to `200`)

Returns `page`

```javascript
await puppeteerT4IT.goto(page, baseUrl);
```

### Timeout execution for specified amount of milliseconds

You can pass following arguments into this function:

* `ms` time in ms to wait

Returns empty Promise

```javascript
await puppeteerT4IT.sleep(500);
```

### Make screenshots from web page

You can pass following arguments into this function:

* `page` Puppeteer [page](#Get-new-page-instance) object
* `width` Screen width
* `height` Screen height
* `pageName` Web page name under which a screenshot will be stored
* `deviceScaleFactor` Specify device scale factor (Defaults to `1`)
* `isMobile` isMobile Whether the meta viewport tag is taken into account (defaults to `false`)
* `delay` Time in ms to wait after reload (defaults to `0`)
* `beforeAction` Function to be called before screenshot is made (defaults to `null`)
* `beforeActionParameters` beforeAction parameters (defaults to `[]`)

Returns empty Promise

```javascript
await puppeteerT4IT.screenshotPage(page, 360, 640, 'frontpage', 1, true, delay, clickOnButton, ['.navigation-element--class']);
```

### Make screenshots in most popular screen resolutions

You can pass following arguments into this function:

* `page` Puppeteer [page](#Get-new-page-instance) object
* `name` Web page name under which screenshots will be stored
* `delay` Time in ms to wait after reload (defaults to `0`)
* `beforeAction` Function to be called before screenshot is made (defaults to `null`)
* `beforeActionParameters` beforeAction parameters (defaults to `[]`)

Returns empty Promise

```javascript
await puppeteerT4IT.screenshotMultipleResolutions(page, 'projects', 4000);
```

### Click on element

You can pass following arguments into this function:

* `page` Puppeteer [page](#Get-new-page-instance) object
* `querySelector` Page element selector
* `delay` Time in ms to wait after reload (defaults to `0`)

Returns empty Promise

```javascript
await puppeteerT4IT.click(page, '.class', 500);
```

### Tap on element

You can pass following arguments into this function:

* `page` Puppeteer [page](#Get-new-page-instance) object
* `querySelector` Page element selector
* `delay` Time in ms to wait after reload (defaults to `0`)

Returns empty Promise

```javascript
await puppeteerT4IT.tapOnElement(page, '.class', 500);
```

### Get href link from element

You can pass following arguments into this function:

* `page` Puppeteer [page](#Get-new-page-instance) object
* `querySelector` Page element selector

Returns link href from element

```javascript
await puppeteerT4IT.getLinkByQuerySelector(page, '.class');
```

### Get href link hostname from element

You can pass following arguments into this function:

* `page` Puppeteer [page](#Get-new-page-instance) object
* `querySelector` Page element selector

It returns link hostname from element.

```javascript
await puppeteerT4IT.getHostnameByQuerySelector(page, '.class');
```

### Get value from element

You can pass following arguments into this function:

* `page` Puppeteer [page](#Get-new-page-instance) object
* `querySelector` Page element selector

Returns value of element

```javascript
await puppeteerT4IT.getValueByQuerySelector(page, '.class');
```

## License

This project is licensed under the MIT License - see the [LICENSE](LICENSE.md) file for details.

## Changelog

Project updates are tracked in a [CHANGELOG](CHANGELOG.md) file


CHANGELOG.md
# CHANGELOG

## 2.0.0 (2023-01-09)

- **breaking** Moved Puppeteer to be just dev dependency so user can install appropriate version of puppeteer or puppeteer-core by himself
- Updated dependencies
- Updated code style

## 1.8.0

- Updated Puppeteer to version 10.2.0
- Updated dependencies
- Fixed lint errors

## 1.7.1

- Fixed lint errors

## 1.7.0

- Updated Puppeteer to version 9.1.1
- Updated dependencies

## 1.6.0

- Updated Puppeteer to version 3.0.4
- Updated dependencies

## 1.5.0

- Updated Puppeteer to version 2.1.1
- Updated dependencies
- Replaced `eslint-config-xo` in favour of `eslint-config-touch4it`

## 1.4.0

- Updated Puppeteer to version 1.20.0
- Updated dependencies

## 1.3.0

- Updated Puppeteer to version 1.18.0
- Updated dependencies

## 1.2.0

- Updated Puppeteer to version 1.17.0
- Updated dependencies

## 1.1.3

- Fixed screenshot clipping

## 1.1.0

- Updated Puppeteer to version 1.13.0
- Added function to get Puppeteer object

## 1.0.4

- Added npm keywords

## 1.0.3

- Updated readme
- Fixed status code assertion variables order

## 1.0.2

- Updated readme
- Renamed screenshot function

## 1.0.1

- Updated readme
- Cleaned package.json
- Renamed variable to match coding style

## 1.0.0

- Initial commit


src/index.js
const assert = require('node:assert');
const index = require('puppeteer');
const fs = require('fs-extra');

module.exports = {
  STATUS_OK: 200,
  STATUS_CREATED: 201,
  STATUS_BAD_REQUEST: 400,
  STATUS_UNAUTHORIZED: 401,
  STATUS_FORBIDDEN: 403,
  STATUS_NOT_FOUND: 404,
  STATUS_SERVER_ERROR: 500,

  TIMEOUT: 30_000,
  SCREENSHOT_DIRECTORY: 'screenshots/',

  /**
   * Initialize puppeteer wrapper
   *
   * @param {number} timeout Request timeout in ms
   * @param {string} screenshotDirectory Path of the directory where screenshots will be stored
   * @returns {void} Does not return any data
   */
  init(timeout = 30_000, screenshotDirectory = 'screenshots/') {
    module.exports.TIMEOUT = timeout;
    module.exports.SCREENSHOT_DIRECTORY = screenshotDirectory;
  },

  /**
   * Launch new browser instance
   *
   * @param {Object} options Browser launch options
   * @returns {Promise<Puppeteer.Browser>} Browser instance
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-puppeteerlaunchoptions
   */
  async launch(options = {}) {
    if (typeof options !== 'object' || options === null) {
      options = {};
    }

    if (!Array.isArray(options.args)) {
      options.args = [];
    }

    options.args.push('--no-sandbox', '--disable-setuid-sandbox', '--disable-dev-shm-usage');

    const browser = await index.launch(options);

    // Browser error handling
    browser.on('disconnected', () => {
      console.log('Chromium has been closed');
    });

    return browser;
  },

  /**
   * Get Puppeteer object
   *
   * @returns {Puppeteer} Puppeteer required library
   */
  getPuppeteer: () => index,

  /**
   * Get new page instance
   *
   * @param {Puppeteer.Browser} browser Browser instance
   * @param {Object?} authenticationData Object containing "username" and "password" keys for basic auth
   * @returns {Promise<Puppeteer.Page>} New page instance
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-browsernewpage
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pageauthenticatecredentials
   */
  async getPage(browser, authenticationData = null) {
    const page = await browser.newPage();

    // Failed request handling
    page.on('requestfailed', request => {
      if (request.failure().errorText !== 'net::ERR_ABORTED') {
        console.error(`Request failed: ${request.url()} ${request.failure().errorText}`);
      }
    });

    // Page error handling
    page.on('error', error => {
      console.error('page on error', error);
    });

    if (authenticationData) {
      await page.authenticate(authenticationData);
    }

    // Content
    await page.setViewport({
      width: 1920,
      height: 1080,
      deviceScaleFactor: 1,
    });

    return page;
  },

  /**
   * Navigate to web page and test status code
   *
   * @param {Puppeteer.Page} page Browser page object
   * @param {string} url URL to be redirected to
   * @param {number} expectedStatusCode awaited HTTP response status code
   * @returns {Promise<Puppeteer.Page>} Browser page
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pagegotourl-options
   * @see https://www.npmjs.com/package/assert#assertequalactual-expected-message
   */
  async goto(page, url, expectedStatusCode = 200) {
    console.log(`Goto ${url}`);

    const response = await page.goto(url, {
      timeout: module.exports.TIMEOUT,
      waitUntil: 'domcontentloaded',
    });

    const receivedStatusCode = response.status();

    if (!response) {
      throw new Error(`Response empty for "${url}"`);
    }

    assert.equal(`${receivedStatusCode}`, `${expectedStatusCode}`, `Wrong status code for "${url}"`);

    return page;
  },

  /**
   * Timeout execution for specified amount of milliseconds
   *
   * @param {int} ms - Time in ms
   * @returns {Promise<void>} Resolved promise when finished
   */
  // eslint-disable-next-line no-promise-executor-return
  sleep: ms => new Promise(resolve => setTimeout(resolve, ms)),

  /**
   * Make screenshots from web page
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {int} width Screen width
   * @param {int} height Screen height
   * @param {string} pageName Web page name
   * @param {number} deviceScaleFactor Specify device scale factor (can be thought of as dpr). Defaults to 1
   * @param {boolean} isMobile Whether the meta viewport tag is taken into account. Defaults to false
   * @param {number} delay Time in ms to wait after reload
   * @param {function} beforeAction Function to be called before screenshot is made
   * @param {array} beforeActionParameters beforeAction parameters
   * @returns {Promise<void>} Resolved promise when finished
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pagesetviewportviewport
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pagereloadoptions
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pagescreenshotoptions
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pageselector
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-elementhandledispose
   */
  async screenshotPage(page, width, height, pageName, deviceScaleFactor = 1, isMobile = false, delay = 0, beforeAction = null, beforeActionParameters = []) {
    try {
      const pathName = `${pageName}-${(`${width}`).padStart(4, '0')}x${(`${height}`).padStart(4, '0')}at${deviceScaleFactor}`;

      console.log(`Screenshotting ’${pathName}’`);

      deviceScaleFactor = Number.parseInt(deviceScaleFactor, 10);

      if (Number.isNaN(deviceScaleFactor) || deviceScaleFactor < 1) {
        throw new Error('device scale factor incorrect');
      }

      await page.setViewport({
        width,
        height,
        deviceScaleFactor,
        isMobile,
      });

      try {
        await page.reload({
          timeout: module.exports.TIMEOUT,
          waitUntil: 'domcontentloaded',
        });
      // eslint-disable-next-line unicorn/prefer-optional-catch-binding
      } catch (_) {
        console.log(`Failed reload: ${pathName}`);
        return;
      }

      // Wait before making screenshot
      if (Number(delay) > 0) {
        await module.exports.sleep(delay);
      }

      // Optionally call method before screenshotting
      if (beforeAction) {
        await beforeAction(...beforeActionParameters);
      }

      // Viewport screenshot
      await page.screenshot({
        path: `${module.exports.SCREENSHOT_DIRECTORY + pageName}/${pathName}-viewport.jpg`,
        fullPage: false,
        type: 'jpeg',
      });

      // Full page screenshot
      const bodyHandle = await page.$('body');
      const boundingBox = await bodyHandle.boundingBox();

      let boundingBoxWidth = Number.parseInt(boundingBox.width, 10);
      boundingBoxWidth = (Number.isNaN(boundingBoxWidth)) ? 1 : boundingBoxWidth;
      boundingBoxWidth = (boundingBoxWidth < 1) ? 1 : boundingBoxWidth;

      let boundingBoxHeight = Number.parseInt(boundingBox.height, 10);
      boundingBoxHeight = (Number.isNaN(boundingBoxHeight)) ? 1 : boundingBoxHeight;
      boundingBoxHeight = (boundingBoxHeight < 1) ? 1 : boundingBoxHeight;

      await page.screenshot({
        path: `${module.exports.SCREENSHOT_DIRECTORY + pageName}/${pathName}-fullpage.jpg`,
        clip: {
          x: 0,
          y: 0,
          width: boundingBoxWidth,
          height: boundingBoxHeight,
        },
        type: 'jpeg',
      });

      await bodyHandle.dispose();

      return;
    } catch (error) {
      console.log('screenshotPage', error);
      throw error;
    }
  },

  /**
   * Make screenshots in most popular screen resolutions
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {string} name Web page name
   * @param {number} delay Time in ms to wait after reload
   * @param {function} beforeAction Function to be called before screenshot is made
   * @param {array} beforeActionParameters beforeAction parameters
   * @returns {Promise<void>} Resolved promise when finished
   *
   * @see https://github.com/jprichardson/node-fs-extra/blob/HEAD/docs/ensureDir.md
   */
  async screenshotMultipleResolutions(page, name, delay = 0, beforeAction = null, beforeActionParameters = []) {
    try {
      await fs.ensureDir(module.exports.SCREENSHOT_DIRECTORY + name);

      await module.exports.screenshotPage(page, 360, 640, name, 1, true, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotPage(page, 360, 640, name, 3, true, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotPage(page, 640, 360, name, 1, true, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotPage(page, 1366, 768, name, 1, false, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotPage(page, 1920, 1080, name, 1, false, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotPage(page, 375, 667, name, 1, true, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotPage(page, 375, 667, name, 2, true, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotPage(page, 667, 375, name, 1, true, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotPage(page, 1440, 900, name, 1, false, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotPage(page, 1280, 800, name, 1, false, delay, beforeAction, beforeActionParameters);
    } catch (error) {
      console.log('screenshotMultipleResolutions', error);
      throw error;
    }
  },

  /**
   * Click on element
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {string} querySelector Page element selector
   * @param {number} delayAfter Time in ms to wait after click
   * @returns {Promise<void>} Resolved promise when finished
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pageclickselector-options
   */
  async clickOnElement(page, querySelector, delayAfter = 0) {
    try {
      await page.click(querySelector);
    // eslint-disable-next-line unicorn/prefer-optional-catch-binding
    } catch (_) {
      console.log(`Element not found or hidden "${querySelector}"`);
    }

    return module.exports.sleep(delayAfter);
  },

  /**
   * Click on element
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {string} querySelector Page element selector
   * @param {number} delayAfter Time in ms to wait after click
   * @returns {Promise<void>} Resolved promise when finished
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pagetapselector
   */
  async tapOnElement(page, querySelector, delayAfter = 0) {
    try {
      await page.tap(querySelector);
    // eslint-disable-next-line unicorn/prefer-optional-catch-binding
    } catch (_) {
      console.log(`Element not found or hidden "${querySelector}"`);
    }

    return module.exports.sleep(delayAfter);
  },

  /**
   * Get href link from <a> element
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {string} querySelectorString DOM selector of target element
   * @returns {Promise<string?>} Link href from element
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pageevalselector-pagefunction-args-1
   */
  getLinkByQuerySelector: (page, querySelectorString) => page.$eval(querySelectorString, link => link.href),

  /**
   * Get href link hostname from <a> element
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {string} querySelectorString DOM selector of target element
   * @returns {Promise<string?>} Link hostname from element
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pageevalselector-pagefunction-args-1
   */
  getHostnameByQuerySelector: (page, querySelectorString) => page.$eval(querySelectorString, link => link.hostname),

  /**
   * Get value of an element
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {string} querySelectorString DOM selector of target element
   * @returns {Promise<string?>} Value of element
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pageevalselector-pagefunction-args-1
   */
  getValueByQuerySelector: (page, querySelectorString) => page.$eval(querySelectorString, link => link.innerHTML),
};


src/puppeteer.js
/* eslint-disable max-len, no-console */

const puppeteer = require('puppeteer');
const assert = require('assert');
const fs = require('fs-extra');

module.exports = {
  STATUS_OK: 200,
  STATUS_CREATED: 201,
  STATUS_BAD_REQUEST: 400,
  STATUS_UNAUTHORIZED: 401,
  STATUS_FORBIDDEN: 403,
  STATUS_NOT_FOUND: 404,
  STATUS_SERVER_ERROR: 500,

  TIMEOUT: 30000,
  SCREENSHOT_DIRECTORY: 'screenshots/',

  /** 
  Initialize puppeteer wrapper
  * @param {number} specify timeout in ms 
  * @return {string} name of the directory where screenshots will be stored
  */
  init: (timeout = 30000, screenshot_directory = 'screenshots/') => {
    module.exports.TIMEOUT = timeout;
    module.exports.SCREENSHOT_DIRECTORY = screenshot_directory;
  },

  /**
   * Launch new browser instance
   *
   * @param {Object} options Browser launch options
   * @return {Promise<Puppeteer.Browser>} Browser instance
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-puppeteerlaunchoptions
   */
  launch: async (options = {}) => {
    if (typeof options !== 'object' || options === null) {
      options = {};
    }

    if (!Array.isArray(options.args)) {
      options.args = [];
    }

    options.args.push('--no-sandbox');
    options.args.push('--disable-setuid-sandbox');
    options.args.push('--disable-dev-shm-usage');

    const browser = await puppeteer.launch(options);

    // Browser error handling
    browser.on('disconnected', () => {
      console.log('Chromium has been closed');
    });

    return browser;
  },

  /**
   * Get new page instance
   *
   * @param {Puppeteer.Browser} browser Browser instance
   * @param {Object?} authenticationData Object containing "username" and "password" keys for basic auth
   * @return {Promise<Puppeteer.Page>} New page instance
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-browsernewpage
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pageauthenticatecredentials
   */
  getPage: async (browser, authenticationData = null) => {
    const page = await browser.newPage();

    // Failed request handling
    page.on('requestfailed', request => {
      if (request.failure().errorText !== 'net::ERR_ABORTED') {
        console.error(`Request failed: ${request.url()} ${request.failure().errorText}`);
      }
    });

    // Page error handling
    page.on('error', err => {
      console.error('page on error', err);
    });

    if (authenticationData) {
      await page.authenticate(authenticationData);
    }

    // Content
    await page.setViewport({
      width: 1920,
      height: 1080,
      deviceScaleFactor: 1
    });

    return page;
  },

  /**
   * Navigate to web page and test status code
   *
   * @param {Puppeteer.Page} page Browser page object
   * @param {string} url URL to be redirected to
   * @param {number} statusCode awaited HTTP response status code
   * @return {Promise<Puppeteer.Page>} Browser page
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pagegotourl-options
   * @see https://www.npmjs.com/package/assert#assertequalactual-expected-message
   */
  goto: async (page, url, statusCode = 200) => {
    console.log(`Goto ${url}`);

    let response;
    let status;

    try {
      response = await page.goto(url, {
        timeout: module.exports.TIMEOUT,
        waitUntil: 'domcontentloaded'
      });

      status = response.status();
    } catch (err) {
      console.error('goto', 'page.goto()', err);
      return Promise.reject(err);
    }

    if (!response) {
      console.error(`Response empty for "${url}"`);
      return Promise.reject(new Error(`Response empty for "${url}"`));
    }

    assert.equal(statusCode, `${status}`, `Wrong status code for "${url}"`);

    return Promise.resolve(page);
  },

  /**
   * Timeout execution for specified amount of milliseconds
   *
   * @param {int} ms - Time in ms
   * @return {Promise<void>} Resolved promise when finished
   */
  sleep: ms => new Promise(resolve => setTimeout(resolve, ms)),

  /**
   * Make screenshots from web page
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {int} width Screen width
   * @param {int} height Screen height
   * @param {string} pageName Web page name
   * @param {number} deviceScaleFactor Specify device scale factor (can be thought of as dpr). Defaults to 1
   * @param {boolean} isMobile Whether the meta viewport tag is taken into account. Defaults to false
   * @param {number} delay Time in ms to wait after reload
   * @param {function} beforeAction Function to be called before screenshot is made
   * @param {array} beforeActionParameters beforeAction parameters
   * @return {Promise<void>} Resolved promise when finished
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pagesetviewportviewport
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pagereloadoptions
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pagescreenshotoptions
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pageselector
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-elementhandledispose
   */
  screenshotTest: async (page, width, height, pageName, deviceScaleFactor = 1, isMobile = false, delay = 0, beforeAction = null, beforeActionParameters = []) => {
    try {
      const pathName = `${pageName}-${(`${width}`).padStart(4, '0')}x${(`${height}`).padStart(4, '0')}at${deviceScaleFactor}`;

      console.log(`Screenshotting '${pathName}'`);

      deviceScaleFactor = parseInt(deviceScaleFactor, 10);

      if (Number.isNaN(deviceScaleFactor) || deviceScaleFactor < 1) {
        const err = new Error('device scale factor incorrect');
        return Promise.reject(err);
      }

      await page.setViewport({
        width,
        height,
        deviceScaleFactor,
        isMobile
      });

      try {
        await page.reload({
          timeout: module.exports.TIMEOUT,
          waitUntil: 'domcontentloaded'
        });
      } catch (err) {
        console.log(`Failed reload: ${pathName}`);
        return Promise.resolve();
      }

      // Wait before making screenshot
      if (Number(delay) > 0) {
        await module.exports.sleep(delay);
      }

      // Optionally call method before screenshotting
      if (beforeAction) {
        await beforeAction(...beforeActionParameters);
      }

      // Viewport screenshot
      await page.screenshot({
        path: `${module.exports.SCREENSHOT_DIRECTORY + pageName}/${pathName}-viewport.jpg`,
        fullPage: false,
        type: 'jpeg'
      });

      // Full page screenshot
      const bodyHandle = await page.$('body');
      const boundingBox = await bodyHandle.boundingBox();

      await page.screenshot({
        path: `${module.exports.SCREENSHOT_DIRECTORY + pageName}/${pathName}-fullpage.jpg`,
        clip: {
          x: 0,
          y: 0,
          width: parseInt(boundingBox.width, 10),
          height: parseInt(boundingBox.height, 10)
        },
        type: 'jpeg'
      });

      await bodyHandle.dispose();

      return Promise.resolve();
    } catch (err) {
      console.log('screenshotTest', err);
      throw err;
    }
  },

  /**
   * Make screenshots in most popular screen resolutions
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {string} name Web page name
   * @param {number} delay Time in ms to wait after reload
   * @param {function} beforeAction Function to be called before screenshot is made
   * @param {array} beforeActionParameters beforeAction parameters
   * @return {Promise<void>} Resolved promise when finished
   *
   * @see https://github.com/jprichardson/node-fs-extra/blob/HEAD/docs/ensureDir.md
   */
  screenshotMultipleResolutions: async (page, name, delay = 0, beforeAction = null, beforeActionParameters = []) => {
    try {
      await fs.ensureDir(module.exports.SCREENSHOT_DIRECTORY + name);

      await module.exports.screenshotTest(page, 360, 640, name, 1, true, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotTest(page, 360, 640, name, 3, true, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotTest(page, 640, 360, name, 1, true, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotTest(page, 1366, 768, name, 1, false, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotTest(page, 1920, 1080, name, 1, false, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotTest(page, 375, 667, name, 1, true, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotTest(page, 375, 667, name, 2, true, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotTest(page, 667, 375, name, 1, true, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotTest(page, 1440, 900, name, 1, false, delay, beforeAction, beforeActionParameters);
      await module.exports.screenshotTest(page, 1280, 800, name, 1, false, delay, beforeAction, beforeActionParameters);
    } catch (err) {
      console.log('screenshotMultipleResolutions', err);
      throw err;
    }

    return Promise.resolve();
  },

  /**
   * Click on element
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {string} querySelector Page element selector
   * @param {number} delayAfter Time in ms to wait after click
   * @return {Promise<void>} Resolved promise when finished
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pageclickselector-options
   */
  clickOnElement: async (page, querySelector, delayAfter = 0) => {
    try {
      await page.click(querySelector);
    } catch (err) {
      console.log(`Element not found or hidden "${querySelector}"`);
    }

    return module.exports.sleep(delayAfter);
  },

  /**
   * Click on element
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {string} querySelector Page element selector
   * @param {number} delayAfter Time in ms to wait after click
   * @return {Promise<void>} Resolved promise when finished
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pagetapselector
   */
  tapOnElement: async (page, querySelector, delayAfter = 0) => {
    try {
      await page.tap(querySelector);
    } catch (err) {
      console.log(`Element not found or hidden "${querySelector}"`);
    }

    return module.exports.sleep(delayAfter);
  },

  /**
   * Get href link from <a> element
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {string} querySelectorString DOM selector of target element
   * @return {Promise<string?>} Link href from element
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pageevalselector-pagefunction-args-1
   */
  getLinkByQuerySelector: (page, querySelectorString) => page.$eval(querySelectorString, link => link.href),

  /**
   * Get href link hostname from <a> element
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {string} querySelectorString DOM selector of target element
   * @return {Promise<string?>} Link hostname from element
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pageevalselector-pagefunction-args-1
   */
  getHostnameByQuerySelector: (page, querySelectorString) => page.$eval(querySelectorString, link => link.hostname),

  /**
   * Get value of an element
   *
   * @param {Puppeteer.Page} page Puppeteer page object
   * @param {string} querySelectorString DOM selector of target element
   * @return {Promise<string?>} Value of element
   *
   * @see https://pptr.dev/#?product=Puppeteer&version=v1.8.0&show=api-pageevalselector-pagefunction-args-1
   */
  getValueByQuerySelector: (page, querySelectorString) => page.$eval(querySelectorString, link => link.innerHTML)
};
